module/zfs/abd.c

```diff
@@ -284,6 +345,100 @@ abd_alloc_sametype(abd_t *sabd, size_t size)
 	}
 }
 
+
+/*
+ * Create Gang ABD that will be the head of a list of ABD's. This is used
+ * to "chain" scatter/gather lists together when constructing aggregated
+ * IO's. To free this abd, abd_free() must be called.
+ */
+abd_t *
+abd_alloc_gang_abd(void)
+{
+	abd_t *abd;
+
+	abd = abd_alloc_struct(0);
+	abd->abd_flags = ABD_FLAG_GANG | ABD_FLAG_OWNER;
+	abd->abd_size = 0;
+	abd->abd_parent = NULL;
+	list_create(&ABD_GANG(abd).abd_gang_chain,
+	    sizeof (abd_t), offsetof(abd_t, abd_gang_link));
+	zfs_refcount_create(&abd->abd_children);
+	return (abd);
+}
+
+/*
+ * Add a child ABD to a Gang ABD's chained list.
+ */
+void
+abd_gang_add(abd_t *pabd, abd_t *cabd, boolean_t free_on_free)
+{
+	ASSERT(abd_is_gang_abd(pabd));
+	abd_t *child_abd = NULL;
+
+	mutex_enter(&cabd->abd_mtx);
+	if (list_link_active(&cabd->abd_gang_link)) {
+		/*
+		 * If the child ABD is already part of another
+		 * Gang ABD then we must allocate a new
+		 * ABD to use a seperate link. We mark the newly
+		 * allocated ABD with ABD_FLAG_GANG_FREE, before
+		 * adding it to the Gang ABD's list, to make the
+		 * Gang ABD is aware that it is it's responsibility
+		 * to call abd_put(). We use abd_get_offset() in order
+		 * to just allocate a new ABD but avoid copying the
+		 * data over into the newly allocated ABD.
+		 *
+		 * Cases where an ABD may be part of multiple
+		 * Gang ABD's are ditto blocks and when
```

Comment:
how about something like: `An ABD may becompe part of multiple gang ABD's.  For example, when writing ditto blocks, the same abd is used to write to 2 or 3 locations with 2 or 3 zio_t's.  Each of the zio's may be aggregated with different adjacent zio's.  zio aggregation uses gang zio's, so the single abd can become part of multiple gang zio's.`

Reply:
I update this comment.